autoprecompiles/src/powdr.rs

```diff
@@ -95,52 +94,63 @@ impl<'a, T: Clone + Ord + std::fmt::Display + 'a, E: AllChildren<AlgebraicExpres
     }
 }
 
-pub fn reassign_ids<T: FieldElement>(
-    mut machine: SymbolicMachine<T>,
-    mut curr_id: u64,
+/// Globalizes the references in the machine by appending a suffix to their names
+/// and offsetting their IDs to start from `curr_id`.
+/// Returns:
+/// - The updated `next_global_id`.
+/// - The substitutions, mapping the local reference IDs to the global ones.
+/// - The updated machine with globalized references.
+pub fn globalize_references<T: FieldElement>(
+    machine: SymbolicMachine<T>,
+    mut next_global_id: u64,
     suffix: usize,
 ) -> (u64, Vec<u64>, SymbolicMachine<T>) {
-    // Build a mapping from local columns to global columns
-    let subs: BTreeMap<AlgebraicReference, AlgebraicReference> = machine
+    let unique_reference_ids = machine.unique_references().map(|r| r.id).collect_vec();
+    let machine_size = unique_reference_ids.len() as u64;
+    assert_eq!(
+        *unique_reference_ids.iter().max().unwrap(),
+        machine_size - 1,
+        "The reference ids must be contiguous"
+    );
+
+    let machine = globalize_reference_names(machine, suffix);
+    let machine = offset_reference_ids(machine, next_global_id);
+
+    let subs = (next_global_id..(next_global_id + machine_size)).collect::<Vec<_>>();
+    next_global_id += machine_size;
+    (next_global_id, subs, machine)
+}
+
+/// Globalizes the names of references in the machine by appending a suffix.
+fn globalize_reference_names<T: FieldElement>(
+    mut machine: SymbolicMachine<T>,
+    suffix: usize,
+) -> SymbolicMachine<T> {
+    // Allocate a new string for each *unique* reference in the machine
+    let globalized_name = |name| Arc::new(format!("{name}_{suffix}"));
+    let name_by_id = machine
         .unique_references()
-        // zip with increasing ids, mutating curr_id
-        .zip(from_fn(|| {
-            let id = curr_id;
```

Comment:
We previously assigned IDs by their order of appearance, now they would be ordered as before. This is very likely the cause of the changes in the snapshot tests.